common/src/main/java/org/conscrypt/HpkeSuite.java

```diff
@@ -0,0 +1,340 @@
+        return AEAD.CHACHA20POLY1305;
+      default:
+        throw new IllegalArgumentException("AEAD " + aead + " not supported.");
+    }
+  }
+
+  enum KEM {
+    DHKEM_X25519_HKDF_SHA256(
+        /* id= */ 0x0020, /* encLength= */ 32, /* pkLength= */ 32, /* skLength= */ 32);
+
+    private final int id;
+    private final int encLength;
+    private final int pkLength;
+    private final int skLength;
+
+    KEM(int id, int encLength, int pkLength, int skLength) {
+      this.id = id;
+      this.encLength = encLength;
+      this.pkLength = pkLength;
+      this.skLength = skLength;
+    }
+
+    /**
+     * KEM id
+     *
+     * @return kem id
+     * @see <a
+     *         href="https://www.rfc-editor.org/rfc/rfc9180.html#name-key-encapsulation-mechanism">KEM
+     *         ids</a>
+     */
+    int getId() {
+      return id;
+    }
+
+    /**
+     * The length in bytes of an encapsulated key produced by this KEM.
+     *
+     * @return encapsulated key size in bytes
+     */
+    int getEncLength() {
+      return encLength;
+    }
+
+    /**
+     * The length in bytes of an encoded public key for this KEM.
+     *
+     * @return public key size in bytes
+     */
+    int getPkLength() {
+      return pkLength;
+    }
+
+    /**
+     * The length in bytes of an encoded private key for this KEM.
+     *
+     * @return private key size in bytes
+     */
+    int getSkLength() {
+      return skLength;
+    }
+
+    /**
+     * Validates the enc size in bytes matches the {@link KEM} spec.
+     *
+     * @param enc encapsulated key produced by the kem
+     * @see <a
+     *         href="https://www.rfc-editor.org/rfc/rfc9180.html#name-key-encapsulation-mechanism">expected
+     *         enc size</a>
+     */
+    void validateEncLength(byte[] enc) {
+      Preconditions.checkNotNull(enc, "enc");
+      final int expectedLength = this.getEncLength();
+      if (enc.length != expectedLength) {
+        throw new IllegalArgumentException(
+            "Expected enc length of " + expectedLength + ", but was " + enc.length);
+      }
+    }
+
+    /**
+     * Validates the public key size in bytes to match the {@link KEM} public key spec and returns
+     * the raw bytes.
+     *
+     * @param publicKey alias pk
+     * @return key in its raw format
+     * @see <a
+     *         href="https://www.rfc-editor.org/rfc/rfc9180.html#name-algorithm-identifiers">expected
+     *         pk size</a>
+     */
+    byte[] validatePublicKeyLengthAndGetRawKey(PublicKey publicKey) {
+      Preconditions.checkNotNull(publicKey, "publicKey");
+      if (!(publicKey instanceof OpenSSLX25519PublicKey)) {
+        throw new IllegalArgumentException(
+            "Public key algorithm " + publicKey.getAlgorithm() + " is not supported");
+      }
+
+      final byte[] key = ((OpenSSLX25519PublicKey) publicKey).getU();
+      final int expectedLength = this.getPkLength();
+      if (key.length != expectedLength) {
```

Comment:
It's a little surprising to me that this is even possible; I'd have expected this to be captured in the type. Ditto below. Though I see the `byte[]` constructor doesn't check, so that's a little odd. I dunno, I'll defer to @prbprbprb here. (If the type guaranteed the length, or is almost always going to be correct anyway, I would suggest we leave out the length validation since BoringSSL will do it (and, in other KEMs, more) anyway.) Just one will thing to worry about.

Reply:
The type *should* indeed guarantee this, and it tries to but as we've discussed the X509/PKCS8 parsing isn't 100% correct.  Let's assume here that the key types already enforce this and I'll take a separate action to fix them.

Reply:
Oops, and yes, the `byte[]` constructors need fixing too.